mountpoint-s3/src/prefetch/parquet_prefetch.rs

```diff
@@ -0,0 +1,181 @@
+    let parquet_magic_len = 8; // Represents the 4 bit "size of metadata" + 4 bit "magic number" (https://parquet.apache.org/docs/file-format/)
+                               // check file is large enough to hold footer
+    let file_size = chunk_reader.len();
+    if file_size < (parquet_magic_len as u64) {
+        return Err(ParquetError::General(
+            "Invalid Parquet file. Size is smaller than footer".to_string(),
+        ));
+    }
+
+    let mut footer = [0_u8; 8];
+    chunk_reader.get_read(file_size - 8, 8)?.read_exact(&mut footer)?;
+
+    let metadata_len = decode_footer(&footer)?;
+    let footer_metadata_len = parquet_magic_len + metadata_len;
+
+    if footer_metadata_len > file_size as usize {
+        return Err(ParquetError::General(
+            "Invalid Parquet file. Reported metadata length is shorter than expected".to_string(),
+        ));
+    }
+
+    let metadata = chunk_reader.get_bytes(file_size - footer_metadata_len as u64, metadata_len)?;
+    Ok(metadata)
+}
+
+struct ParquetS3ChunkReader<Client: ObjectClient> {
+    client: Arc<Client>,
+    bucket: String,
+    key: String,
+    etag: ETag,
+    size: u64,
+}
+
+impl<Client: ObjectClient + Send + Sync + 'static> ChunkReader for ParquetS3ChunkReader<Client> {
+    type T = S3ReadSeek<Client>;
+
+    fn get_read(&self, start: u64, length: usize) -> Result<Self::T, ParquetError> {
+        Ok(S3ReadSeek {
+            client: self.client.clone(),
+            bucket: self.bucket.clone(),
+            key: self.key.clone(),
+            etag: self.etag.clone(),
+            position: start,
+            size: self.size,
+            length,
+        })
+    }
+}
+
+struct S3ReadSeek<Client: ObjectClient> {
+    client: Arc<Client>,
+    bucket: String,
+    key: String,
+    etag: ETag,
+    position: u64,
+    size: u64,
+    length: usize,
+}
+
+impl<Client: ObjectClient + Send + Sync + 'static> Read for S3ReadSeek<Client> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        trace!("Entered here for reader s3seek??");
+        let range_start = self.position;
+        let range_end = (self.position + self.length.min(buf.len()) as u64).min(self.size);
+        let range: std::ops::Range<u64> = range_start..range_end;
+
+        let span = debug_span!("read", range = ?range);
+        let _enter = span.enter();
+
+        if range.is_empty() {
+            return Ok(0);
+        }
+        let expected_size = range.end - range.start;
+        let mut data = Vec::with_capacity(expected_size as usize);
+
+        let runtime = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let read_len = runtime.block_on(async {
+            let get_object_result = self
+                .client
+                .get_object(&self.bucket, &self.key, Some(range), Some(self.etag.clone()))
```

Comment:
making an s3 request per `read` may be too expensive, do we know how much reads `parquet` library typically emit for a single file? (irrelevant if we download the footer in advance and then serve reads from the in-memory buffer)

Reply:
Usually we'll have only one pair of requests to the metadata per file handle (PrefetchGetObject instance), so wouldn't be too expensive. And as discussed, normally if the metadata field is empty, that means the first request by a query engine for that file handle would typically be for the metadata, so we can serve straight from the cache.